src/cmd/serial.js

```diff
@@ -379,69 +379,21 @@ module.exports = class SerialCommand {
 		]);
 	}
 
-	flashDevice(binary, { port, yes }){
-		let device;
-
-		return Promise.resolve()
-			.then(() => {
-				if (!yes){
-					return this._promptForListeningMode();
-				}
-			})
-			.then(() => this.whatSerialPortDidYouMean(port, true))
-			.then(_device => {
-				device = _device;
-				if (!device){
-					throw new VError('No serial port identified');
-				}
-				if (device.type === 'Core'){
-					throw new VError('Serial flashing is not supported on the Core');
-				}
-
-				//only match against knownApp if file is not found
-				let stats;
-				try {
-					stats = fs.statSync(binary);
-				} catch (ex){
-					// file does not exist
-					const specsByProduct = _.keyBy(specs, 'productName');
-					const productSpecs = specsByProduct[device.type];
-					binary = productSpecs && productSpecs.knownApps[binary];
-					if (binary === undefined){
-						throw new VError('File does not exist and no known app found');
-					} else {
-						return;
-					}
-				}
+	async flashDevice(binary, { port, applicationOnly=false }) {
+		const device = await this.whatSerialPortDidYouMean(port, true);
+		if (!device) {
+			throw new VError('No serial port identified');
+		}
 
-				if (!stats.isFile()){
-					throw new VError('You cannot flash a directory over USB');
-				}
-			})
-			.then(() => {
-				const serialPort = new SerialPort(device.port, {
-					baudRate: 28800,
-					autoOpen: false
-				});
+		// TODO: Check if this still holds true or if we even care
+		if ((device.specs && device.specs.name === 'core') || device.type === 'Core') {
+			throw new VError('Serial flashing is not supported on the Core');
+		}
 
-				const closePort = () => {
-					if (serialPort.isOpen){
-						serialPort.close();
-					}
-					process.exit(0);
-				};
-				process.on('SIGINT', closePort);
-				process.on('SIGTERM', closePort);
+		const deviceId = device.deviceId;
 
-				const ymodem = new YModem(serialPort, { debug: true });
-				return ymodem.send(binary);
-			})
-			.then(() => {
-				console.log('\nFlash success!');
-			})
-			.catch(err => {
-				throw new VError(ensureError(err), 'Error writing firmware');
-			});
+		const flashCmdInstance = new FlashCommand();
+		await flashCmdInstance.flashLocal({ files: [deviceId, binary], applicationOnly });
 	}
```

Comment:
The `this.whatSerialPortDidYouMean(port, true);` ideally returns a device object with the following structure, IF the provided port number matches with the connected device's port number. If it doesn't match, then `device` object will look like this and this `device: { port: '/dev/cu.usbmodem144101', type: '' }` function will crash because it does not have a deviceId. (line 393). Previously, we would open `SerialPort` and send binary anyway on that serial port. Is this a breaking change?

```
Device:  {
  port: '/dev/tty.usbmodem144101',
  type: 'Boron',
  deviceId: 'e00fce68b329990d97948c8a',
  specs: {
    name: 'boron',
    productName: 'Boron',
    productId: 13,
    generation: 3,
    features: [ 'cellular', 'udp', 'mesh', 'ble' ],
    defaultProtocol: 'udp',
    serial: { vid: '2b04', pid: 'c00d' },
    knownApps: {
      tinker: '/Users/keerthyamisagadda/code/particle-cli/assets/knownApps/boron/tinker/tinker-0.8.0-rc.27-boron.bin'
    },
    tcpServerKey: {
      address: 2082,
      size: 512,
      format: 'der',
      alt: 1,
      alg: 'rsa',
      addressOffset: 384,
      portOffset: 450
    },
    udpServerKey: {
      address: 3298,
      size: 320,
      format: 'der',
      alt: 1,
      alg: 'ec',
      addressOffset: 192,
      portOffset: 258
    },
    tcpPrivateKey: { address: 34, size: 612, format: 'der', alt: 1, alg: 'rsa' },
    udpPrivateKey: { address: 3106, size: 192, format: 'der', alt: 1, alg: 'ec' },
    factoryReset: { address: '0x00200000', alt: 2 },
    formerUserPart: { address: '0xd4000', alt: 0, size: 4096 }
  }
}
```

Reply:
I added a check for device.deviceId in the method. That should be good enough.